databroker/experimental/server_ext.py

```diff
@@ -0,0 +1,503 @@
+
+    def __init__(self, collection, directory, doc):
+        self.collection = collection
+        self.directory = directory
+        self.doc = Document(**doc)
+        self.array_adapter = None
+        if self.doc.data_url is not None:
+            path = self.doc.data_url.path
+            if platform == "win32" and path[0] == "/":
+                path = path[1:]
+
+            file = h5py.File(path)
+            dataset = file["data"]
+            self.array_adapter = ArrayAdapter(dask.array.from_array(dataset))
+        elif self.doc.data_blob is not None:
+            self.array_adapter = ArrayAdapter(dask.array.from_array(self.doc.data_blob))
+
+    @property
+    def structure(self):
+        return ArrayStructure.from_json(self.doc.structure)
+
+    @property
+    def metadata(self):
+        return self.doc.metadata
+
+    @array_raise_if_inactive
+    def read(self, *args, **kwargs):
+        return self.array_adapter.read(*args, **kwargs)
+
+    @array_raise_if_inactive
+    def read_block(self, *args, **kwargs):
+        return self.array_adapter.read_block(*args, **kwargs)
+
+    def microstructure(self):
+        return self.array_adapter.microstructure()
+
+    def macrostructure(self):
+        return self.array_adapter.macrostructure()
+
+    def put_data(self, body):
+        # Organize files into subdirectories with the first two
+        # charcters of the uid to avoid one giant directory.
+        path = self.directory / self.doc.uid[:2] / self.doc.uid
+        path.parent.mkdir(parents=True, exist_ok=True)
+        array = numpy.frombuffer(
+            body, dtype=self.doc.structure.micro.to_numpy_dtype()
+        ).reshape(self.doc.structure.macro.shape)
+        with h5py.File(path, "w") as file:
+            file.create_dataset("data", data=array)
+        self.collection.update_one(
+            {"uid": self.doc.uid},
+            {
+                "$set": {
+                    "data_url": f"file://localhost/{str(path).replace(os.sep, '/')}"
+                }
+            },
+        )
+
+
+class WritingDataFrameAdapter:
+    structure_family = "dataframe"
+
+    def __init__(self, collection, directory, doc):
+        self.collection = collection
+        self.directory = directory
+        self.doc = Document(**doc)
+        self.dataframe_adapter = None
+
+        if self.doc.data_url is not None:
+            path = self.doc.data_url.path
+            if platform == "win32" and path[0] == "/":
+                path = path[1:]
+
+            self.dataframe_adapter = DataFrameAdapter(
+                dask.dataframe.from_pandas(
+                    pandas.read_parquet(path),
+                    npartitions=self.doc.structure.macro.npartitions,
+                )
+            )
+
+        elif self.doc.data_blob is not None:
+            self.dataframe_adapter = DataFrameAdapter(
+                dask.dataframe.from_pandas(
+                    deserialize_arrow(base64.b64decode(self.doc.data_blob)),
```

Comment:
Reminder to remove base64 here (and from the tiled PR).

Reply:
why is this resolved? I don't think it makes sense to base64 encode in mongo.

Reply:
Also if the document model supports `data_blob` should there be a way to make the writing adapters use this?

Reply:
> why is this resolved? I don't think it makes sense to base64 encode in mongo.

I will remove this part. It was not checked by the tests because `data_blob` is not being implemented yet. We left a space for it in the Document for our next steps of development. Dan was thinking on letting the server decide whether writing the data in the document in `data_blob` or save it in disc and link it with `data_url`, depending on the size of the data